text/3268-repr-scalable.md

```diff
@@ -0,0 +1,191 @@
+writing I believe cranelift doesn't support scalable vectors ([current proposal](https://github.com/bytecodealliance/rfcs/pull/19)),
+and the GCC backend is not mature enough to be thinking about this.
+
+Most of the complexity of SVE will be handled by LLVM and the `vscale` modifier that is applied to vector types. Therefore
+changes for this should be fairly minimal for Rust. From the LLVM side this is as simple as calling `LLVMScalableVectorType`
+rather than `LLVMVectorType`.
+
+For a Scalable Vector Type LLVM takes the form `<vscale x elements x type>`.
+* `elements` multiplied by sizeof(`type`) gives the smallest allowed register size and the increment size.
+* `vscale` is a run time constant that is used to determine the actual vector register size.
+
+For example, with Arm SVE the scalable vector register (Z register) size has to
+be a multiple of 128 bits and a power of 2 (via a retrospective change to the
+architecture), therefore for `f32`, `elements` would always be four. At run time
+`vscale` could be 1, 2, 4, 8, 16 which would give register sizes of 128, 256,
+512, 1024 and 2048. While SVE now has the power of 2 restriction, `vscale` could
+be any value providing it gives a legal vector register size for the
+architecture.
+
+The scalable representation accepts the number of `elements` rather than the compiler calculating it, which serves
+two purposes. The first being that it removes the need for the compiler to know about the user defined types and how to calculate
+the required `element` count. The second being that some of these scalable types can have different element counts. For instance,
+the predicates used in SVE have different element counts in LLVM depending on the types they are a predicate for.
+
+As mentioned previously `vscale` is a runtime constant. With SVE the vector length can be changed at runtime (e.g. by a
+[prctl()](https://www.kernel.org/doc/Documentation/arm64/sve.txt) call in Linux). However, since this would require a change
+to `vscale`, this is considered undefined behaviour in Rust. This is consistent with C and C++ implementations.
+
+## Unsized rules
+These types aren't `Sized`, but they need to exist in local variables, and we
+need to be able to pass them to, and return them from functions. This means
+adding an exception to the rules around returning unsized types in Rust. There
+are also some traits (`Copy`) that have a bound on being `Sized`.
+
+We will implement `Copy` for these types within the compiler, without having to
+implement the traits when the types are defined.
+
+This RFC also changes the rules so that function return values can be `Copy` or
+`Sized` (or both). Once returning of unsized is allowed this part of the rule
+would be superseded by that mechanism. It's worth noting that, if any other
+types are created that are `Copy` but not `Sized` this rule would apply to
+those.
```

Comment:
Remember that Rust has generics, so I can e.g. write a function `fn foo<T: Copy>(x: &T) -> T`. The RFC seems to say this is allowed, because the return type is `Copy`. But for most types `T` and most ABIs this can't be implemented.

You can't just say in a sentence that you allow unsized return values. That's a major language feature that needs significant design work on its own.

I think what you actually want is some extremely special cases where specifically these scalable vector types are allowed as return values, but in a non-compositional way. There is no precedent for anything like this in Rust so it needs to be fairly carefully described and discussed.